File: data/derived_cot/rq3_prompting/results/translation/r1/java_to_cpp/code_output/MetricsCalculator.cpp
#include <iostream>
#include <vector>

class MetricsCalculator {
public:
    int truePositives;
    int falsePositives;
    int falseNegatives;
    int trueNegatives;

    MetricsCalculator() : truePositives(0), falsePositives(0), falseNegatives(0), trueNegatives(0) {}

    void update(const std::vector<int>& predictedLabels, const std::vector<int>& trueLabels) {
        for (std::size_t i = 0; i < predictedLabels.size(); ++i) {
            if (predictedLabels[i] == 1 && trueLabels[i] == 1) {
                ++truePositives;
            } else if (predictedLabels[i] == 1 && trueLabels[i] == 0) {
                ++falsePositives;
            } else if (predictedLabels[i] == 0 && trueLabels[i] == 1) {
                ++falseNegatives;
            } else if (predictedLabels[i] == 0 && trueLabels[i] == 0) {
                ++trueNegatives;
            }
        }
    }

    double precision(const std::vector<int>& predictedLabels, const std::vector<int>& trueLabels) {
        update(predictedLabels, trueLabels);
        if (truePositives + falsePositives == 0) {
            return 0.0;
        }
        return static_cast<double>(truePositives) / (truePositives + falsePositives);
    }

    double recall(const std::vector<int>& predictedLabels, const std::vector<int>& trueLabels) {
        update(predictedLabels, trueLabels);
        if (truePositives + falseNegatives == 0) {
            return 0.0;
        }
        return static_cast<double>(truePositives) / (truePositives + falseNegatives);
    }

    double f1Score(const std::vector<int>& predictedLabels, const std::vector<int>& trueLabels) {
        update(predictedLabels, trueLabels);
        double prec = precision(predictedLabels, trueLabels);
        double rec = recall(predictedLabels, trueLabels);
        if (prec + rec == 0.0) {
            return 0.0;
        }
        return (2 * prec * rec) / (prec + rec);
    }

    double accuracy(const std::vector<int>& predictedLabels, const std::vector<int>& trueLabels) {
        update(predictedLabels, trueLabels);
        int total = truePositives + trueNegatives + falsePositives + falseNegatives;
        if (total == 0) {
            return 0.0;
        }
        return static_cast<double>(truePositives + trueNegatives) / total;
    }
};

int main() {
    MetricsCalculator mc;
    std::vector<int> predictedLabels = {1, 1, 0, 0};
    std::vector<int> trueLabels = {1, 0, 0, 1};
    std::cout << mc.precision(predictedLabels, trueLabels) << std::endl;
    std::cout << mc.recall(predictedLabels, trueLabels) << std::endl;
    std::cout << mc.f1Score(predictedLabels, trueLabels) << std::endl;
    std::cout << mc.accuracy(predictedLabels, trueLabels) << std::endl;
    return 0;
}

File: data/derived_cot/rq2_eval/translation/qwen/_converted_code/java_to_py/MetricsCalculator.py
class MetricsCalculator:

    def __init__(self):
        self.truePositives = 0
        self.falsePositives = 0
        self.falseNegatives = 0
        self.trueNegatives = 0

    def update(self, predictedLabels, trueLabels):
        for i in range(len(predictedLabels)):
            if predictedLabels[i] == 1 and trueLabels[i] == 1:
                self.truePositives += 1
            elif predictedLabels[i] == 1 and trueLabels[i] == 0:
                self.falsePositives += 1
            elif predictedLabels[i] == 0 and trueLabels[i] == 1:
                self.falseNegatives += 1
            elif predictedLabels[i] == 0 and trueLabels[i] == 0:
                self.trueNegatives += 1

    def precision(self, predictedLabels, trueLabels):
        self.update(predictedLabels, trueLabels)
        total_positive = self.truePositives + self.falsePositives
        if total_positive == 0:
            return 0.0
        return float(self.truePositives) / total_positive

    def recall(self, predictedLabels, trueLabels):
        self.update(predictedLabels, trueLabels)
        total_relevant = self.truePositives + self.falseNegatives
        if total_relevant == 0:
            return 0.0
        return float(self.truePositives) / total_relevant

    def f1_score(self, predictedLabels, trueLabels):
        self.update(predictedLabels, trueLabels)
        precision_val = self.precision(predictedLabels, trueLabels)
        recall_val = self.recall(predictedLabels, trueLabels)
        if precision_val + recall_val == 0.0:
            return 0.0
        return 2 * (precision_val * recall_val) / (precision_val + recall_val)

    def accuracy(self, predictedLabels, trueLabels):
        self.update(predictedLabels, trueLabels)
        total = self.truePositives + self.trueNegatives + self.falsePositives + self.falseNegatives
        if total == 0:
            return 0.0
        return float(self.truePositives + self.trueNegatives) / total